moveit_core/trajectory_processing/src/iterative_spline_parameterization.cpp

```diff
@@ -0,0 +1,510 @@
+
+/* Author: Ken Anderson */
+
+/*
+  Code to time parameterize a trajectory into a cubic spline
+  while respecting velocity, acceleration, and jerk constraints.
+*/
+
+#include <moveit/trajectory_processing/iterative_spline_parameterization.h>
+#include <moveit_msgs/JointLimits.h>
+#include <console_bridge/console.h>
+#include <moveit/robot_state/conversions.h>
+#include <vector>
+
+#define VLIMIT 1.0  // default
+#define ALIMIT 3.0  // default
+#define JLIMIT 9.0  // default
+
+/*
+  Comment-out to disable jerk limits.
+  This would be the case if you want to allow bang-bang style control,
+  where the controller can go from min acceleration to max acceleration
+  instantaneously.  In other words, the acceleration curve is discontinuous.
+  Since the acceleration can change instantly, matching the initial and final
+  acceleration is unneccessary, as is adding and moving the two extra points.
+  This will get the shortest paths and should be comparable to 'optimal'
+  methods that have a trapazoidal velocity curve.
+
+  However, fitting a cubic spline creates a trajectory with a continuous
+  acceleration curve.
+  If you are using this technique, then you probably don't want rapid changes
+  in acceleration like a bang-bang controller.
+  This means that jerks limits need to be enforced.
+  The downside is that if the limits are low, you will notice visible oscilation
+  around what you may expect as the 'optimal' velocity curve.
+  Increasing the jerk limits will reduce oscilation and speed up execution.
+  By default, this should be ENABLED.
+*/
+#define ENABLE_JERK
+
+namespace trajectory_processing
+{
+static void fit_cubic_spline(const int n, const double dt[], const double x[], double x1[], double x2[]);
+static void adjust_two_positions(const int n, const double dt[], double x[], double x1[], double x2[],
+                                 const double x2_i, const double x2_f);
+static void init_times(const int n, double dt[], const double x[], const double max_velocity);
+static int fit_spline_and_adjust_times(const int n, double dt[], const double x[], double x1[], double x2[],
+                                       const double vlimit, const double alimit, const double jlimit,
+                                       const double tfactor);
+
+// The path of a single joint: positions, velocities, and accelerations
+struct SingleJointTrajectory
+{
+  std::vector<double> positions;  // joint's position at time[x]
+  std::vector<double> velocities;
+  std::vector<double> accelerations;
+  double initial_velocity;
+  double final_velocity;
+  double initial_acceleration;
+  double final_acceleration;
+  double max_velocity;
+  double max_acceleration;
+  double max_jerk;
+};
+
+IterativeSplineParameterization::IterativeSplineParameterization(double max_time_change_per_it, bool add_points)
+  : max_time_change_per_it_(1.0 + max_time_change_per_it), add_points_(add_points)
```

Comment:
to make it a multiplicative factor.  I changed the member variable name to reflect this.